### PR TITLE
Add more tests for collections

### PR DESCRIPTION
**Pull request by layomia**
Additional tests for collections, given the changes in https://github.com/dotnet/corefx/pull/41482.
- Fixes https://github.com/dotnet/corefx/issues/40657
- Partly addresses https://github.com/dotnet/corefx/issues/40370

### CODE REVIEW

File: src/System.Text.Json/tests/Serialization/Value.ReadTests.ConcurrentCollections.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            ConcurrentQueue<string> qc = JsonSerializer.Deserialize<ConcurrentQueue<string>>(@"[""1""]");
+            Assert.Equal(1, qc.Count);
+            qc.TryPeek(out string val);
+            Assert.Equal("1", val);
+
+            ConcurrentStack<string> qs = JsonSerializer.Deserialize<ConcurrentStack<string>>(@"[""1""]");
+            Assert.Equal(1, qs.Count);
+            qc.TryPeek(out val);
+            Assert.Equal("1", val);
+        }
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/Value.ReadTests.ConcurrentCollections.cs]: `ConcurrentBag<T>` as well?

File: src/System.Text.Json/tests/Serialization/Value.ReadTests.ConcurrentCollections.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+        [Fact]
+        public static void Read_ConcurrentCollection_Throws()
+        {
+            // Not supported. Not IList, and we don't detect the add method for this collection.
+            Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<BlockingCollection<string>>(@"[""1""]"));
+
+            // Not supported. Not IList, and we don't detect the add method for this collection.
+            Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<ConcurrentBag<string>>(@"[""1""]"));
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/Value.ReadTests.ConcurrentCollections.cs]: Why isn't `ConcurrentBag<T>` supported but `ConcurrentQueue<T>` and `ConcurrentStack<T>` are?  All three implement the same two interfaces: `IProducerConsumerCollection<T>` and `IReadOnlyCollection<T>`.
https://github.com/dotnet/corefx/blob/007f4456374c582ebc7a7a678072935a2560eee5/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs#L40
https://github.com/dotnet/corefx/blob/007f4456374c582ebc7a7a678072935a2560eee5/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs#L33
https://github.com/dotnet/corefx/blob/007f4456374c582ebc7a7a678072935a2560eee5/src/Common/src/CoreLib/System/Collections/Concurrent/ConcurrentQueue.cs#L22

File: src/System.Text.Json/tests/Serialization/Value.ReadTests.SpecializedCollections.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            ListDictionary ld = JsonSerializer.Deserialize<ListDictionary>(@"{""key"":""value""}");
+            Assert.Equal(1, ld.Count);
+            Assert.Equal("value", ((JsonElement)ld["key"]).GetString());
+        }
+
+        [Fact]
+        public static void Read_SpecializedCollection_Throws()
+        {
+            // Add method for this collection only accepts strings, even though it only implements IList which usually
+            // indicates that the element type is typeof(object).
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/Value.ReadTests.SpecializedCollections.cs]: This seems like a strange limitation; we support non-generic collections... why is it a problem for the Add method to support a more derived type?

File: src/System.Text.Json/tests/Serialization/Value.WriteTests.ConcurrentCollections.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+namespace System.Text.Json.Serialization.Tests
+{
+    public static partial class ValueTests
+    {
+        [Fact]
+        public static void Write_ConcurrentCollection()
+        {
+            Assert.Equal(@"[""1""]", JsonSerializer.Serialize(new BlockingCollection<string> { "1" }));
+
+            Assert.Equal(@"[""1""]", JsonSerializer.Serialize(new ConcurrentBag<string> { "1" }));
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/Value.WriteTests.ConcurrentCollections.cs]: So we support serializing some types but not deserializing them?

File: src/System.Text.Json/tests/Serialization/Value.ReadTests.ConcurrentCollections.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+using Xunit;
+
+namespace System.Text.Json.Serialization.Tests
+{
+    public static partial class ValueTests
+    {
+        [Fact]
+        public static void Read_ConcurrentCollection()
+        {
+            ConcurrentDictionary<string, string> cd = JsonSerializer.Deserialize<ConcurrentDictionary<string, string>>(@"{""key"":""value""}");
```
Comment (steveharter) [inline src/System.Text.Json/tests/Serialization/Value.ReadTests.ConcurrentCollections.cs]: IMO having a single test for each of these doing both deserialize + serialize as well as verifying rount-tripping capability (e.g. an extra serialize or deserialize) would be more readable and better test the functionality.

I realize some existing tests separate the "read" and "write", but going forward I don't think that is better.

**Comment by steveharter** [comment]
As part of the "auditing" [PR ](#40657)it would be goodness to have that PR display what collection types are supported and what is not -- e.g. a list of all collection types and interfaces, can they be derived from and whether they support both serialize and deserialize.

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

**Comment by ahsonkhan** [comment]
Looks like this PR has some questions/discussion worth following up on. We can consider closing this and re-opening it in dotnet/runtime, when ready.

**Comment by layomia** [comment]
> Looks like this PR has some questions/discussion worth following up on. We can consider closing this and re-opening it in dotnet/runtime, when ready.

Closing for this reason. Will address feedback shortly in dotnet/runtime.

**Pull request closed without merging**